File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    public void solve(char[][] board) {
        int n = board.length;
        int m = board[0].length;
        int[][] visited = new int[n][m];
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                visited[i][j] = 0;
            }
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if((i==0 || j == 0 || i == n-1 || j == m-1) && board[i][j] == 'O' && visited[i][j] == 0){
                    dfs(visited,board,i,j);
                }
            }
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(visited[i][j] == 0 && board[i][j]=='O'){
                    board[i][j] = 'X';
                }
            }
        }
    }
     static void dfs(int[][] visited,char[][] board,int row,int col){
            int n = visited.length;
            int m = visited[0].length;
            visited[row][col] = 1;
            int[][] direc = {{1,0},{-1,0},{0,-1},{0,1}};
            for(int i = 0;i<direc.length;i++){
                int nr = row+direc[i][0];
                int nc = col+direc[i][1];
                if(nr>=0 && nr<n && nc>=0 && nc<m && board[nr][nc] == 'O' && visited[nr][nc]==0){
                    dfs(visited,board,nr,nc);
                }
            }
        } 
}

File: 0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        n = len(board)
        m = len(board[0])
        visited = [[0]*m for _ in range(n)]
        def dfs(row,col):
            visited[row][col] = 1
            direc = [[-1,0],[1,0],[0,-1],[0,1]]
            for dr,dc in direc:
                nr,nc = row+dr,col+dc
                if 0<=nr<n and 0<=nc<m and board[nr][nc] == 'O' and visited[nr][nc] == 0:
                    dfs(nr,nc)
        for i in range(n):
            for j in range(m):
                if (i == 0 or j == 0 or i == n-1 or j == m-1) and board[i][j] == 'O' and visited[i][j] == 0:
                    dfs(i,j)

        for i in range(n):
            for j in range(m):
                if visited[i][j] == 0 and board[i][j] == "O":
                    board[i][j] = "X"
        return board 